Libraries/LibGfx/Painter.cpp

```diff
@@ -1137,4 +1149,157 @@ void Painter::stroke_path(const Path& path, Color color, int thickness)
     }
 }
 
+//#define FILL_PATH_DEBUG
+
+void Painter::fill_path(Path& path, Color color, WindingRule winding_rule)
+{
+    const auto& segments = path.split_lines();
+
+    if (segments.size() == 0)
+        return;
+
+    Vector<Path::LineSegment> active_list;
+    active_list.ensure_capacity(segments.size());
+
+    // first, grab the segments for the very first scanline
+    auto first_y = segments.first().maximum_y;
+    auto last_y = segments.last().minimum_y;
+    auto scanline = first_y;
+
+    size_t last_active_segment { 0 };
+
+    for (auto& segment : segments) {
+        if (segment.maximum_y != scanline)
+            break;
+        active_list.append(segment);
+        ++last_active_segment;
+    }
+
+    auto is_inside_shape = [winding_rule](int winding_number) {
+        if (winding_rule == WindingRule::Nonzero)
+            return winding_number != 0;
+
+        if (winding_rule == WindingRule::EvenOdd)
+            return winding_number % 2 == 0;
+
+        ASSERT_NOT_REACHED();
+    };
+
+    auto increment_winding = [winding_rule](int& winding_number, const Point& from, const Point& to) {
+        if (winding_rule == WindingRule::EvenOdd) {
+            ++winding_number;
+            return;
+        }
+
+        if (winding_rule == WindingRule::Nonzero) {
+            if (from.dy_relative_to(to) < 0)
+                ++winding_number;
+            else
+                --winding_number;
+            return;
+        }
+
+        ASSERT_NOT_REACHED();
+    };
+
+    while (scanline >= last_y) {
+        if (active_list.size()) {
+            // sort the active list by 'x' from right to left
+            qsort(active_list.data(), active_list.size(), sizeof(Path::LineSegment), [](const void* d0, const void* d1) -> int {
```

Comment:
Can we use `AK::quick_sort` instead here?

Reply:
Huh. I didn't know we _has_ an `AK::quick_sort`. we absolutely can.
(there's also one in Path.cpp)